### ISSUE TITLE
When on a branch, pushing should issue a git command using the branch name instead of HEAD

### ISSUE THREAD

**Issue by SkySkimmer** [magit/magit#3847]
Typically I change some stuff on a branch `foo`, ask magit to push (magit-push-current-to-pushremote) and without waiting switch to another branch `bar`.
This means `git … push -v origin HEAD` races with `git … checkout bar` (apparently git waits for some network operation / long enough for user action before deciding what it's pushing).

**Comment by kyleam** [comment]
I take it that you actually hit into this in the wild, which seems plausible based on stepping through builtin/push.c.  To push the branch specifically, I think we'd just need the following change, though perhaps there's a reason for using HEAD that I'm not thinking of.

```diff
diff --git a/lisp/magit-push.el b/lisp/magit-push.el
index bf73248de..138dc655e 100644
--- a/lisp/magit-push.el
+++ b/lisp/magit-push.el
@@ -86,10 +86,10 @@ (define-suffix-command magit-push-current-to-pushremote (args)
   :if 'magit-get-current-branch
   :description 'magit-push--pushbranch-description
   (interactive (list (magit-push-arguments)))
-  (pcase-let ((`(,_ ,remote)
+  (pcase-let ((`(,branch ,remote)
                (magit--select-push-remote "push there")))
     (run-hooks 'magit-credential-hook)
-    (magit-run-git-async "push" "-v" args remote "HEAD")))
+    (magit-run-git-async "push" "-v" args remote branch)))

 (defun magit-push--pushbranch-description ()
   (let* ((branch (magit-get-current-branch))
```

**Comment by tarsius** [comment]
I have choosen `HEAD` over `<branch>` because its documentation sounds more like what we want to do. Except for the obvious difference these commands might not differ at all, in which case `<branch>` would be preferable. But to know for sure I will have to think of what the edge-cases are and then test how they affect the behavior of these commands.

```
       git push origin master
           Find a ref that matches master in the source repository (most
           likely, it would find refs/heads/master), and update the same ref
           (e.g.  refs/heads/master) in origin repository with it. If master
           did not exist remotely, it would be created.

       git push origin HEAD
           A handy way to push the current branch to the same name on the
           remote.
```

According to this, the latter pushes a branch to a branch, while the former pushes something somewhere, usually a branch to a branch, but anything is possible.

> (apparently git waits for some network operation / long enough for user action before deciding what it's pushing)

I would argue that this is a bug in Git, which should be reported.

**Comment by kyleam** [comment]
> Except for the obvious difference these commands might not differ at all, in which case `<branch>` would be preferable. But to know for sure I will have to think of what the edge-cases are and then test how they affect the behavior of these commands.

FWIW I don't spot any scenarios or configuration that would result in `git push origin refs/heads/CURRENT` behaving differently than `git push origin HEAD`.  I think the "most likely, it would find refs/heads/master" bit in the description of `git push origin master` is covering the weird case where there is a "master" tag but no branch (if both existed, the command would fail entirely) rather than alluding to any configuration that could change the behavior.

**Comment by tarsius** [comment]
After doing some reading and experiments I have concluded that `HEAD`, `<name>` and `refs/heads/<name>` always do the same, except for the one difference @kyleam mentioned (`<name>` fails when there is a tag by the same name), and that we should therefore use the third.

I verified that:
* `remote.<name>.fetch` and `remote.<name>.pull` have no effect,
* local ambiguous non-tag refs have no effect (e.g. `refs/<name>`, and
* remote ambiguous refs (including tags) have no effect (e.g. `refs/<name>` or `refs/tags/<name>`).

Also there is no way to have a local branch that is not in the `refs/heads/*` namespace. There are many ways to configure a repository in very strange ways, but this particular footgun is not available. So prefixing with `refs/heads/` is not a problem.
